src/hangar/external/manage_plugins.py

```diff
@@ -0,0 +1,219 @@
+from collections import defaultdict
+import configparser
+import os
+from glob import glob
+from warnings import warn
+from typing import Callable, Tuple
+
+
+# TODO: test cases
+
+
+def _clear_plugins():
+    """
+    Clear the plugin state to the default, i.e., where no plugins are loaded
+    """
+    plugin_details_store.clear()
+    plugin_provides.clear()
+    plugin_module_name.clear()
+    plugin_meta_data.clear()
+    _plugin_store.clear()
+
+
+def _scan_plugins():
+    """
+    Scan the plugins directory for .ini files and parse them
+    to autofill storage
+    """
+    pd = os.path.dirname(__file__)
+    config_files = glob(os.path.join(pd, '_plugins', '*.ini'))
+
+    for filename in config_files:
+        name, meta_data = _parse_config_file(filename)
+        plugin_meta_data[name] = meta_data
+
+        for a in meta_data['accepts'].split(','):
+            for p in meta_data['provides'].split(','):
+                accepts = a.strip()
+                provides = p.strip()
+                if provides not in valid_provides:
+                    warn(f"Plugin: {name} wants to provide non-existent: {p}. Ignoring.")
+                    continue
+                plugin_provides[name].append(provides)
+                tmp_acc = plugin_details_store.get(accepts)
+                if tmp_acc:
+                    tmp_prv = tmp_acc.get(provides)
+                    if tmp_prv:
+                        tmp_prv.append(name)
+                    else:
+                        tmp_acc[provides] = [name]
+                else:
+                    plugin_details_store[accepts] = {provides: [name]}
+        plugin_module_name[name] = os.path.basename(filename)[:-4]
+
+
+def _read_preferred_plugins_config():
+    """
+    Populate preferred plugin dict that maps file formats to plugins and methods. This
+    is used to infer which plugin to use at runtime based on file format
+    """
+    parser = configparser.ConfigParser()
+    config_path = os.path.join(os.path.dirname(__file__), 'preferred_plugins.ini')
+    parser.read(config_path)
+    return parser
+
+
+def _write_preferred_plugins_config(parser: configparser.ConfigParser):
```

Comment:
It's not ok to dynamically write source files to a library. This would get way to complex to debug in the future.

Reply:
Right, moving it out.

Reply:
Resolved due to discussion over vid chat. Will be handled by in memory structures to avoid this.